pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -1048,8 +1051,29 @@ protected void handleSend(CommandSend send, ByteBuf headersAndPayload) {
 
         startSendOperation();
 
-        // Persist the message
-        producer.publishMessage(send.getProducerId(), send.getSequenceId(), headersAndPayload, send.getNumMessages());
+        if (isTxn) {
+            final long txnIdMostBits = send.getTxnidMostBits();
+            final long txnIdLeastBits = send.getTxnidLeastBits();
+            final long sequenceId = send.getSequenceId();
+            final long producerId = send.getProducerId();
+            TxnID txnID = new TxnID(txnIdMostBits, txnIdLeastBits);
+            headersAndPayload.retain();
+            producer.getTopic()
+                    .getTxnBuffer(true)
+                    .thenCompose(txnBuffer ->
+                                     txnBuffer.appendBufferToTxn(txnID, sequenceId, headersAndPayload))
+                    .thenAccept(position ->
+                                   ctx.writeAndFlush(
+                                       Commands.newSendReceipt(producerId, sequenceId, ((PositionImpl)position).getLedgerId(), ((PositionImpl)position).getEntryId())))
+                    .exceptionally(err -> {
+                        log.warn("Send transaction [{}] message to transaction buffer failed, {}", txnID, err.getCause().getMessage());
+                        ctx.writeAndFlush(Commands.newSendError(producerId, sequenceId, ServerError.UnknownError, err.getCause().getMessage()));
```

Comment:
why it is `UnknownError`?

Reply:
I'm not sure what's the transaction error should use which `ServerError`, so I use `UnknowError` to represent all transaction errors. Maybe I should add a new ServerError type `TxnError` to represent transaction error? Any suggestions?